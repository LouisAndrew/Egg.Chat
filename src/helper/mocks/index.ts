export { mockChatroom } from './chatroom';
export { mockMessage } from './message';
export { mockUser2, mockUser1 } from './user';
export { mockFirebase } from './firebase';
