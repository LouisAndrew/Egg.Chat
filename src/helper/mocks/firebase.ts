// use just as reference.
// Error message: Jest mock must be an inline function

import { Chatroom, User } from '../schema';
import { mockUser1, mockUser2, mockChatroom } from '.';

const mockUserDb: { [key: string]: User } = {
    '1234': mockUser1,
    '5678': mockUser2,
};

const mockChatroomDb: { [key: string]: Chatroom } = {
    '0972': mockChatroom,
};

export const mockFirebase = () => ({
    firestore: () => ({
        collection: (ref: string) => {
            if (ref === 'user') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockUserDb[id]);
                            }),
                    }),
                };
            } else if (ref === 'chatroom') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockChatroomDb[id]);
                            }),
                    }),
                };
            } else {
                return;
            }
        },
    }),
    auth: () => ({}),
});
