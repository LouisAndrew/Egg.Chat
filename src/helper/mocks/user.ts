import { Status, User } from '../schema';

const CHATROOM_ID = '0972';

const mockUser1: User = {
    uid: '1234',
    displayImage: '',
    displayName: 'John Doe',
    status: Status.ONLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(16023160238),
};

const mockUser2: User = {
    uid: '5678',
    displayImage: '',
    displayName: 'Jane Doe',
    status: Status.OFFLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(1605460238),
};

export { mockUser2, mockUser1 };
