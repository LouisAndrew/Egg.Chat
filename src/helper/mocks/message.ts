import { Message } from '../schema';
import { mockUser1 } from './user';

const mockMessage: Message = {
    msgId: '45725',
    msg: 'Hello, World',
    sentBy: mockUser1.uid,
    sentAt: new Date(Date.parse('10 Oct 2020 00:12:00 GMT')),
};

export { mockMessage };
