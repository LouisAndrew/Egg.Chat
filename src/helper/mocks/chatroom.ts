import { Chatroom } from 'helper/schema';
import { mockUser2, mockUser1 } from './user';
import { mockMessage } from './message';

const mockChatroom: Chatroom = {
    roomId: '0972',
    users: [mockUser1.uid, mockUser2.uid],
    messages: [mockMessage],
};

export { mockChatroom };
