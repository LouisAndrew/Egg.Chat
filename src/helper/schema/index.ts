import { Chatroom } from './chatroom';
import { User, Status } from './user';
import { Message } from './message';

export { Status };
export type { Chatroom, User, Message };
