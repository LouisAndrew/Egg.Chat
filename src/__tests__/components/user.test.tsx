import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
// import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import User from 'components/user';
import { mockUser1 } from 'helper/mocks';
import WithTheme from 'helper/util/with-theme';

describe('User', () => {
    const mockUserExtended = {
        ...mockUser1,
        displayImage: 'http://localhost/src',
    };

    const Element = (
        <WithTheme>
            <User {...mockUserExtended} />
        </WithTheme>
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('Should render the user name, image and status correctly', () => {
        const { getByTestId, getByRole } = render(Element);

        const name = getByTestId('user-name');
        const img = getByRole('img');
        const status = getByTestId('status');

        expect(name).toHaveTextContent(mockUser1.displayName);
        expect((img as HTMLImageElement).src).toBe(
            mockUserExtended.displayImage
        );
        expect(status).toHaveTextContent(mockUser1.status);
    });

    // it('Should render the component in its SMALL variant if variant props is not given.', () => {
    //     const { getByRole } = render(Element);

    //     const img = getByRole('img');

    //     expect(img).toHaveStyle('height: 85px;');
    // });

    // it('Should render the comoponent in its BIG variant if variant props is given as big', () => {
    //     const { getByRole } = render(Element);

    //     const img = getByRole('img');

    //     expect(img).toHaveStyle('height: 100px;');
    // });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
