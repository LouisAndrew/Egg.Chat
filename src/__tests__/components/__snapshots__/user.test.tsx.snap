// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User matches snapshot 1`] = `
<div
  className="css-369q98"
>
  <img
    className="css-9taffg"
    src="http://localhost/src"
  />
  <div
    className="details css-1cc7kk0"
  >
    <h2
      className="user-name css-45gygr"
      data-testid="user-name"
    >
      John Doe
    </h2>
    <div
      className="status css-huhpy"
      data-testid="status"
    >
      <div
        className="css-1jntkoe"
      />
      Online
    </div>
  </div>
</div>
`;
